sdk/opa.go

```diff
@@ -0,0 +1,276 @@
+			return err
+		}
+		opa.configBytes = bs
+		return nil
+	}
+}
+
+// Config sets the configuration to use on the OPA instance.
+func Config(config string) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.configBytes = []byte(config)
+		return nil
+	}
+}
+
+// Paths sets the paths to search for policy, data files and bundles.
+func Paths(paths []string) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.paths = paths
+		return nil
+	}
+}
+
+// Store allows providing storage implementation for OPA.
+func Store(store storage.Store) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.store = store
+		return nil
+	}
+}
+
+// Logger sets the logging implementation to use for SDK integration.
+func Logger(logger logging.Logger) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.logger = logger
+		return nil
+	}
+}
+
+// New returns a new OPA object.
+func New(opts ...func(*OPA) error) (*OPA, error) {
+	opa := &OPA{
+		mtx:                   new(sync.Mutex),
+		preparedQueries:       map[string]*rego.PreparedEvalQuery{},
+		preparedQueriesDoOnce: map[string]*sync.Once{},
+	}
+
+	for _, opt := range opts {
+		if err := opt(opa); err != nil {
+			return nil, err
+		}
+	}
+
+	id, err := uuid.New(rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	var pluginInitializers []func(*plugins.Manager)
+
+	if opa.logger == nil {
+		opa.logger = logging.NewNoOpLogger()
+	}
+	pluginInitializers = append(pluginInitializers, plugins.Logger(opa.logger))
+
+	if len(opa.paths) > 0 {
+		loaded, err := initload.LoadPaths(opa.paths, opa.filter, opa.bundleMode, nil, false)
+		if err != nil {
+			return nil, fmt.Errorf("load error %w", err)
+		}
+		pluginInitializers = append(pluginInitializers, plugins.InitBundles(loaded.Bundles))
+		pluginInitializers = append(pluginInitializers, plugins.InitFiles(loaded.Files))
+	}
+
+	if opa.store == nil {
+		opa.store = inmem.New()
+	}
+
+	opa.manager, err = plugins.New(opa.configBytes, id, opa.store, pluginInitializers...)
+	if err != nil {
+		return nil, err
+	}
+
+	discovery, err := discovery.New(opa.manager)
+	if err != nil {
+		return nil, err
+	}
+
+	opa.manager.RegisterCompilerTrigger(opa.compilerUpdated)
+	opa.manager.Register("discovery", discovery)
+
+	opa.interQueryBuiltinCache = iCache.NewInterQueryCache(opa.manager.InterQueryBuiltinCacheConfig())
+
+	return opa, nil
+}
+
+// Start asynchronously starts the policy engine's plugins that download
+// policies, report status, etc.
+func (opa *OPA) Start(ctx context.Context) error {
+	return opa.manager.Start(ctx)
+}
+
+// Stop asynchronously stops the policy engine's plugins.
+func (opa *OPA) Stop(ctx context.Context) {
+	opa.manager.Stop(ctx)
+}
+
+// Query sends a query to the OPA instance and returns the decision
+func (opa *OPA) Query(ctx context.Context, decisionPath string, input interface{}, opts ...func(*rego.Rego)) (interface{}, error) {
+	m := metrics.New()
+	m.Timer("sdk_query").Start()
+	defer m.Timer("sdl_query").Stop()
```

Comment:
nit: typo here

Reply:
I just realized this indicates a lack of test coverage. We could simply assert the resulting metrics have non-zero values for certain timers of importance (`sdk_query`).

Reply:
Created a decision struct to return to be able to carry metrics back to the caller. Test added for this.